docs/protocol/core/v3.0.rst

```diff
@@ -95,7 +96,27 @@ provide.
 Node names
 ----------
 
-TODO define constraints on node names
+Except for the root node, each node in a hierarchy must have a name,
+which is a string of ASCII characters. To ensure consistent behaviour
+across different storage systems, the following constraints apply to
+node names:
+
+* must not be the empty string ("")
+
+* must consist only of characters in the sets `a-z`, `A-Z`, `0-9`,
+  `-_.`
+
+* must not be a string composed only of period characters, e.g. "." or
+  ".."
+
+* must be at most 255 characters long
+
+Note that node names are used to form storage keys, and that some
+storage systems will perform a case-insensitive comparison of storage
+keys during retrieval. Therefore, within a hierarchy, all nodes within
+a set of sibling nodes must have a name that is unique under
+case-insensitive comparison. E.g., the names "foo" and "FOO" are not
+allowed for sibling nodes.
```

Comment:
This restriction is to deal with the situation where file system storage is used on a platform with a case insensitive file system, e.g., current default on mac I believe.